wpt_common.py

```diff
@@ -141,3 +141,107 @@ def fetch_all_prs():
+    # --format="%(refname:strip=2) %(objectname)" would also include SHA-1
+    output = wpt_git(['tag', '--list', 'merge_pr_*'])
+    return set(pr_number_from_tag(tag) for tag in output.split() if tag)
+
+
+def git_contained_pr(commit):
+    """Returns the number of the latest contained PR of commit using git describe."""
+    try:
+        tag = wpt_git(['describe', '--tags', '--match', 'merge_pr_*', commit])
+        return pr_number_from_tag(tag)
+    except subprocess.CalledProcessError:
+        return None
+
+
+def pr_number(pr):
+    return int(pr['PR'])
+
+
+def pr_date(pr):
+    return dateutil.parser.parse(pr['merged_at'])
+
+
+def get_pr_latencies(prs, events=None, event_sha=None, event_date=None):
+    """For each PR, find the earliest event that included that PR,
+    and calucate the latencies between the PR and the event.
+
+    Args:
+        prs: list of PRs from fetch_all_prs()
+        events: list of events in any format
+        event_sha: function to get a sha from an event
+        event_date: function to get a datetime.datetime from an event
+
+    Returns list of { 'pr': pr, 'event': event, 'latency': latency } objects.
+    """
+
+    # Sort the PRs by merge date and filter out the ones that weren't tagged,
+    # since we don't know at which commit they were merged, and their computed
+    # latency can therefore be higher than the real latency. (Any PR which has
+    # correct merge information via the GitHub API should also have a tag:
+    # https://github.com/foolip/ecosystem-infra-stats/issues/6#issuecomment-375731858)
+    tagged_prs = get_tagged_prs()
+    prs = sorted(filter(lambda pr: pr_number(pr) in tagged_prs, prs),
+                 key=pr_date)
+
+    # We get PR-to-event latencies by the following process:
+    #  1. For each event find the latest contained PR. This is done using git
+    #     describe, and after this point the git commit graph doesn't matter.
+    #     Not all events necessarily have a known contained PR at all.
+    #  2. Reverse to a mapping from PR to event. Some PRs won't be the latest
+    #     contained PR of any event, and if there are multiple events for a
+    #     single PR only the earliest is saved.
+    #  3. Walk the list of PR and associated events backwards, keeping track
+    #     of the earliest event encountered so far. That is the event against
+    #     which the latency for the PR is measured.
+
+    # Step 1.
+    # event_contained_prs is a list of PR numbers, not dicts.
+    event_contained_prs = [git_contained_pr(
+        event_sha(event)) for event in events]
+
+    # earliest_event allows using None as a placeholder for no event.
+    def earliest_event(event1, event2):
+        if event1 is None:
+            return event2
+        if event2 is None:
+            return event1
+        return min(event1, event2, key=event_date)
+
+    # Step 2.
+    # pr_earliest_events is a dict with PR numbers as the key.
+    pr_earliest_events = {}
```

Comment:
Perhaps `earliest_events_by_prs` or `prs_to_earliest_events`?

Reply:
Yeah. This started out as a parallel list to prs and named in analogy with event_contained_prs, but that's no longer the case.

Reply:
Went with `earliest_events_for_pr`